scheduling/taskmetadata.go

```diff
@@ -0,0 +1,533 @@
+package scheduling
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/featureform/fferr"
+	"github.com/featureform/locker"
+	ss "github.com/featureform/storage"
+)
+
+type TaskMetadataList []TaskMetadata
+
+type TaskRunList []TaskRunMetadata
+
+func (trl *TaskRunList) FilterByStatus(status Status) {
+	var newList TaskRunList
+	for _, run := range *trl {
+		if run.Status == status {
+			newList = append(newList, run)
+		}
+	}
+	*trl = newList
+}
+
+type TaskMetadataManager struct {
+	storage ss.MetadataStorage
+}
+
+func NewMemoryTaskMetadataManager() TaskMetadataManager {
+	memoryLocker := locker.MemoryLocker{
+		LockedItems: sync.Map{},
+		Mutex:       &sync.Mutex{},
+	}
+
+	memoryStorage := ss.MemoryStorageImplementation{
+		Storage: make(map[string]string),
+	}
+
+	memoryMetadataStorage := ss.MetadataStorage{
+		Locker:  &memoryLocker,
+		Storage: &memoryStorage,
+	}
+	return TaskMetadataManager{
+		storage: memoryMetadataStorage,
+	}
```

Comment:
what if it's

```
    return TaskMetadataManager{
        storage: ss.NewMemoryMetadataStorage(),
    }
```

and you delegate the creation

Reply:
for learnings, what was the reason for no factory functions?